File: Difficult/239 - SlidingWindowMaximum/Solution.cs
// Question 239. Sliding Window Maximum (https://leetcode-cn.com/problems/sliding-window-maximum/)

public class Solution 
{
    private List<(int, int)> currentWindow;
    private int K;

    public int[] MaxSlidingWindow(int[] nums, int k) 
    {
        if (k == 1) {  return nums; }
        
        this.K = k;
        int N = nums.Count();  
        int[] maxSlidingWindows = new int[N - k + 1];
        this.currentWindow = new List<(int, int)>();
        
        // Initialize the windows.
        int maxOfWindow = this.InitializeWindow(nums);
        maxSlidingWindows[0] = maxOfWindow;

        System.Console.WriteLine("21");
        for (int i = k - 1, windowIndex = 1; i <= N - 1; i++, windowIndex++)
        {
            maxSlidingWindows[windowIndex] = this.UpdateWindowAndGetMaxValue(i - k + 1, i, nums[i]);
        }

        return maxSlidingWindows;
    }

    private int InitializeWindow(int[] nums)
    {
        int maxOfWindow = nums[0];

        for (int i = 0; i < this.K; i++)
        {
            int currentValue = nums[i];
            
            int windowCount = this.currentWindow.Count();
            if (windowCount == 0)
            {
                this.currentWindow.Add((currentValue, i));
            } 
            else 
            {
                for (int existingValueIndex = 0; i < windowCount; i++)
                {
                    (int value, int index) = this.currentWindow[existingValueIndex];
                    if (value < currentValue)
                    {
                        this.currentWindow.Insert(i, (currentValue, i));
                    }
                }

                if (this.currentWindow.Count() == windowCount)
                {
                    this.currentWindow.Add((currentValue, i));
                }
            }

            // this.UpdateWindowAndGetMaxValue(-1, i, currentValue);
            maxOfWindow = Math.Max(maxOfWindow, currentValue);
        }

        return maxOfWindow;
    }

    private int UpdateWindowAndGetMaxValue(int indexToRemove, int indexToAdd, int valueToAdd)
    {
        this.currentWindow = this.currentWindow.Where(windowValue => windowValue.Item1 == indexToRemove).ToList();
        
        for (int i = 0; i < this.currentWindow.Count(); i++)
        {
            System.Console.WriteLine($"i = {i}, this.currentWindow.Count() = {this.currentWindow.Count()}");

            (int value, int index) = this.currentWindow[i];
            if (value < valueToAdd)
            {
                this.currentWindow.Insert(i, (valueToAdd, indexToAdd));
                break;
            }
        }

        (int maxValue, _) = this.currentWindow[0];
        return maxValue;
    }
}

File: Difficult/239 - SlidingWindowMaximum/solution.ts
// Question 239. Sliding Window Maximum (https://leetcode-cn.com/problems/sliding-window-maximum/)


/**
 * 执行用时：356 ms, 在所有 TypeScript 提交中击败了 74.36% 的用户
 * 内存消耗：72.2 MB, 在所有 TypeScript 提交中击败了 41.03% 的用户
 *
 * Acknowledgement: This solution borrows idea from the [official solution](https://leetcode-cn.com/problems/sliding-window-maximum/solution/hua-dong-chuang-kou-zui-da-zhi-by-leetco-ki6m/)
 * 
 * @param {number[]} nums
 * @param {number} k
 * @returns {number[]}
 */
function maxSlidingWindow(nums: number[], k: number): number[] {
  if (k === 1) {
    return nums;
  }

  const N: number = nums.length;

  const q: number[] = [];
  for (let index = 0; index < k; index++) {
    while (q.length && nums[index] >= nums[q[q.length - 1]]) {
      q.pop();
    }
    q.push(index);
  }

  const ans: number[] = [nums[q[0]]];
  for (let index = k; index < N; index++) {
    while (q.length && nums[index] >= nums[q[q.length - 1]]) {
      q.pop();
    }

    q.push(index);

    while (q[0] <= index - k) {
      q.shift();
    }

    ans.push(nums[q[0]]);
  }

  return ans;
}

function maxSlidingWindowTimeoutSolution(nums: number[], k: number): number[] {
  if (k === 1) {
    return nums;
  }

  const N: number = nums.length;

  let priorityQueue: [number, number][] = [];
  let slidingWindow: number[] = [];
  let windowMax: number = nums[0];

  for (let index = 0; index < k; index++) {
    const currentValue: number = nums[index];
    priorityQueue.push([currentValue, index]);
    windowMax = Math.max(windowMax, currentValue);
  }
  slidingWindow.push(windowMax);
  priorityQueue.sort(([aValue, _], [bValue, b]) => aValue - bValue);

  for (let index = k; index < N; index++) {
    const currentValue: number = nums[index];
    priorityQueue = priorityQueue.filter(
      ([_, valueIndex]) => valueIndex !== index - k
    );

    priorityQueue.push([currentValue, index]);
    slidingWindow.push(Math.max(...priorityQueue.map(([value, _]) => value)));
  }

  return slidingWindow;
}
